clippy_lints/src/manual_map.rs

```diff
@@ -176,21 +179,38 @@ impl LateLintPass<'_> for ManualMap {
                     } else {
                         ""
                     };
-                    format!(
-                        "|{}{}| {}",
-                        annotation,
-                        some_binding,
-                        snippet_with_context(cx, some_expr.span, expr_ctxt, "..", &mut app).0
-                    )
-                },
+                    if some_expr.needs_unsafe_block {
+                        format!(
+                            "|{}{}| unsafe {{ {} }}",
+                            annotation,
+                            some_binding,
+                            snippet_with_context(cx, some_expr.expr.span, expr_ctxt, "..", &mut app).0
+                        )
+                    } else {
+                        format!(
+                            "|{}{}| {}",
+                            annotation,
+                            some_binding,
+                            snippet_with_context(cx, some_expr.expr.span, expr_ctxt, "..", &mut app).0
+                        )
+                    }
+                }
             }
         } else if !is_wild_none && explicit_ref.is_none() {
             // TODO: handle explicit reference annotations.
-            format!(
-                "|{}| {}",
-                snippet_with_context(cx, some_pat.span, expr_ctxt, "..", &mut app).0,
-                snippet_with_context(cx, some_expr.span, expr_ctxt, "..", &mut app).0
-            )
+            if some_expr.needs_unsafe_block {
+                format!(
+                    "|{}| unsafe {{ {} }}",
+                    snippet_with_context(cx, some_pat.span, expr_ctxt, "..", &mut app).0,
+                    snippet_with_context(cx, some_expr.expr.span, expr_ctxt, "..", &mut app).0
+                )
+            } else {
+                format!(
+                    "|{}| {}",
+                    snippet_with_context(cx, some_pat.span, expr_ctxt, "..", &mut app).0,
+                    snippet_with_context(cx, some_expr.expr.span, expr_ctxt, "..", &mut app).0
+                )
+            }
```

Comment:
Both `format!` macros use the same arguments, here it would be better to store them in a value beforehand. That avoids duplicated expressions and is better readable IMO :upside_down_face: 

```rs
let closure_expr = snippet_with_context(cx, some_expr.expr.span, expr_ctxt, "..", &mut app).0;
let closure_args = snippet_with_context(cx, some_pat.span, expr_ctxt, "..", &mut app).0;
if some_expr.needs_unsafe_block {
    format!("|{}| unsafe {{ {} }}", closure_args, closure_expr)
} else {
    format!("|{}| {}", closure_args, closure_expr)
}
```